tests/unit/s2n_ktls_test.c

```diff
@@ -38,5 +75,163 @@ int main(int argc, char **argv)
+        };
+
+        /* drain buffer prior to enabling kTLS */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            uint8_t write_byte = 8;
+            uint8_t read_byte = 0;
+            /* write to conn->out buffer and assert error */
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&server_conn->out, &write_byte, 1));
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+            /* drain conn->out buffer and assert success case */
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&server_conn->out, &read_byte, 1));
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_send(server_conn));
+
+            /* write to conn->in buffer and assert error */
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&server_conn->in, &write_byte, 1));
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_recv(server_conn), S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+            /* drain conn->in buffer and assert success case */
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&server_conn->in, &read_byte, 1));
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_recv(server_conn));
+        };
+
+        /* managed_send_io */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            /* expect failure if connection is using custom IO */
+            server_conn->managed_send_io = false;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_KTLS_MANAGED_IO);
+
+            /* expect success if connection is NOT using custom IO */
+            server_conn->managed_send_io = true;
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_send(server_conn));
+        };
+
+        /* managed_recv_io */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            /* recv managed io */
+            server_conn->managed_recv_io = false;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_recv(server_conn), S2N_ERR_KTLS_MANAGED_IO);
+
+            /* expect success if connection is NOT using custom IO */
+            server_conn->managed_recv_io = true;
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_recv(server_conn));
+        };
+
+        /* s2n_ktls_retrieve_file_descriptor */
+        {
```

Comment:
I like that you test this, since you skip the actual configuration of the socket in unit tests!